spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/repository/query/SpannerStatementQueryTests.java

```diff
@@ -0,0 +1,268 @@
+
+		PartTreeSpannerQuery partTreeSpannerQuery = new PartTreeSpannerQuery(Trade.class,
+				queryMethod, spannerOperations, new SpannerMappingContext());
+
+		Object[] params = new Object[] {
+				"BUY",
+				"abcd",
+				"abc123",
+				8.88,
+				3.33,
+				new Trade(),
+				"ignored",
+		};
+
+		partTreeSpannerQuery.execute(params);
+	}
+
+	@Test(expected = UnsupportedOperationException.class)
+	public void deleteTest() {
+
+		QueryMethod queryMethod = mock(QueryMethod.class);
+
+		when(queryMethod.getName()).thenReturn(
+				"deleteTop3DistinctIdActionPriceByActionAndSymbolOrTraderIdAndPriceLessThanOrPriceGreater"
+						+ "ThanEqualAndIdIsNotNullAndTraderIdIsNullOrderByIdDesc");
+
+		SpannerOperations spannerOperations = mock(SpannerOperations.class);
+
+		PartTreeSpannerQuery partTreeSpannerQuery = new PartTreeSpannerQuery(Trade.class,
+				queryMethod, spannerOperations, new SpannerMappingContext());
+
+		Object[] params = new Object[0];
+
+		partTreeSpannerQuery.execute(params);
+	}
+
+	@Test(expected = UnsupportedOperationException.class)
+	public void unSupportedPredicateTest() {
+
+		QueryMethod queryMethod = mock(QueryMethod.class);
+
+		when(queryMethod.getName()).thenReturn("countByTraderIdBetween");
+
+		SpannerOperations spannerOperations = mock(SpannerOperations.class);
+
+		PartTreeSpannerQuery partTreeSpannerQuery = new PartTreeSpannerQuery(Trade.class,
+				queryMethod, spannerOperations, new SpannerMappingContext());
+
+		Object[] params = new Object[0];
+
+		partTreeSpannerQuery.execute(params);
+	}
+
+	@Test
+	public void countTest() {
+
+		QueryMethod queryMethod = mock(QueryMethod.class);
+
+		when(queryMethod.getName()).thenReturn("countByAction");
+
+		SpannerOperations spannerOperations = mock(SpannerOperations.class);
+
+		PartTreeSpannerQuery partTreeSpannerQuery = new PartTreeSpannerQuery(Trade.class,
+				queryMethod, spannerOperations, new SpannerMappingContext());
+
+		Object[] params = new Object[]{
+				"BUY",
+		};
+
+		List<Trade> results = new ArrayList<>();
+		results.add(new Trade());
+
+		when(spannerOperations.find(any(), (Statement) any(), any()))
+				.thenReturn((List) results);
+
+		assertEquals(1, partTreeSpannerQuery.execute(params));
+	}
+
+	@Test
+	public void existsTest() {
+
+		QueryMethod queryMethod = mock(QueryMethod.class);
+
+		when(queryMethod.getName()).thenReturn("existsByAction");
+
+		SpannerOperations spannerOperations = mock(SpannerOperations.class);
+
+		PartTreeSpannerQuery partTreeSpannerQuery = new PartTreeSpannerQuery(Trade.class,
+				queryMethod, spannerOperations, new SpannerMappingContext());
+
+		Object[] params = new Object[]{
+				"BUY",
+		};
+
+		List<Trade> results = new ArrayList<>();
+		results.add(new Trade());
+
+		when(spannerOperations.find(any(), (Statement) any(), any()))
+				.thenReturn((List) results);
+
+		assertTrue((boolean) partTreeSpannerQuery.execute(params));
+	}
+
+	@Test
+	public void notExistsTest() {
```

Comment:
lot of duplication between this and the `existsTest`, can you please refactor?

Reply:
I'm confused between combining and having tests separately... Please change as you see fit.

Reply:
well, those are two really separate things: 
1.) a single testcase should have a single reason to fail (Single responsibility principle) - so when you have a compound test, the same testcase will be broken for a long time during a major refactoring for example. Typically it is easy to write this way the test (especially if you write the tests after the code is written!), but for maintenance it's suboptimal (which naturally comes out if you write code in a test driven way). 
2.) code duplication is most of the time bad - in unit test cases that's why we end up with parameterized test cases. The two methods here do the same thing with a single difference: if the returned array is empty, the result should be `false` otherwise `true`. We can create a helper method that takes a given array, and the expected return value, exercises runs the logic and assertion. We'll still have two separate test cases though at the end!

Reply:
I'm not quite sure where the threshold for "readability"/repetitiveness is. This method here is just a few lines,

Reply:
My threshold is max 4-5 lines. This is 8.

Reply:
fixed. (just so happens my threshold was 9!)

Reply:
@balopat PTAL, i'd like to get this commit in so I can work on that SQL statement annotation and also the paging and sorting repo implementations !